### PR TITLE
Update security configs

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature
/kind cleanup

**What this PR does / why we need it**:
- Drop dependabot config for release branches. We don't need to update dependencies on release branches since dependencies get carried over from main and security patches are not affected via this dependabot config.
  - Moreover, it's a bit of a pain to have to manually update the config file every time we create a new release branch.
- Add dependabot config for /hack/tools/go.mod
- Add dependabot config for dockerfile
- Add documentation on installing pre-commit framework and hooks for the codebase

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Merge https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3703 before merging current PR.

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [x] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nawazkh** [comment]
Waiting for https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3703 to be merged.
/hold

**Comment by nawazkh** [comment]
/hold
I noticed that we added `pre-commit-config.yaml` but have no instructions on setting up in the repo. 
Updating..

**Comment by nawazkh** [comment]
- Running `brew install pre-commit` does take a minute or so to complete. 
- Then the user is expected to run `pre-commit install`
- Then running `git commit` gives the following output
```
❯ git commit
[INFO] Initializing environment for https://github.com/gitleaks/gitleaks.
[INFO] Initializing environment for https://github.com/golangci/golangci-lint.
[INFO] Initializing environment for https://github.com/jumanjihouse/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/pylint-dev/pylint.
[INFO] Installing environment for https://github.com/gitleaks/gitleaks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/golangci/golangci-lint.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Installing environment for https://github.com/pylint-dev/pylint.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
Detect hardcoded secrets.................................................Passed
golangci-lint........................................(no files to check)Skipped
Test shell scripts with shellcheck...................(no files to check)Skipped
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
pylint...............................................(no files to check)Skipped
```
  - running `pre-commit` install the first time does take a minute or two. 
- running `git commit` the next time is fast.
```
❯ git commit -m "amend commits"
Detect hardcoded secrets.................................................Passed
golangci-lint........................................(no files to check)Skipped
Test shell scripts with shellcheck...................(no files to check)Skipped
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
pylint...............................................(no files to check)Skipped
[update_security_configs 24e8c77e] amend commits
 2 files changed, 3 insertions(+), 7 deletions(-)
```

File: .github/workflows/codeql.yml [hunk 1]
```diff
@@ -35,7 +35,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        language: ["go", "python"]
+        language: ["go", "python"] # TODO: are we missing any more languages?
```
[the remark this answers is not in the document]
Comment (nawazkh) [reply .github/workflows/codeql.yml]: I created another PR to add local make targets into pre-commit-config. https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3724
I did not add https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3724 changes to new additions.

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

Dependabot changes look reasonable and the added instructions are good. :+1:

**Pull request merged**